kernel/src/main/java/org/kframework/backend/kore/ModuleToKORE.java

```diff
@@ -499,6 +498,98 @@ private void genUnitAxiom(Production prod, StringBuilder sb) {
         sb.append(") [unit{}()] // left unit\n");
     }
 
+    private void genMapCeilAxioms(Production prod, Collection<Rule> rules) {
+        Sort mapSort = prod.nonterminal(1).sort();
+        scala.collection.Set<Production> mapProds = module.productionsForSort().apply(mapSort.head());
+        Production concatProd = mapProds.find(p -> hasHookValue(p.att(), "MAP.concat")).get();
+        Production elementProd = mapProds.find(p -> hasHookValue(p.att(), "MAP.element")).get();
+        List<K> args = new ArrayList<>();
+        List<K> setArgs = new ArrayList<>();
+        Seq<NonTerminal> nonterminals = elementProd.nonterminals();
+        K restMap = KVariable("Rest", Att.empty().add(Sort.class, mapSort));
+        K restMapSet = KVariable("@Rest", Att.empty().add(Sort.class, mapSort));
```

Comment:
the element variables are used for the in_keys rules. The set variables are used for the ceil rules.

Reply:
I find it difficult to follow how the rules are constructed because the construction of the rules is interleaved in this way, but it's not difficult enough that I'm going to hold up this pull request.

Reply:
I've refactored the code and added comments with the rules. hope it is clearer now.